src/rust/engine/hashing/src/lib.rs

```diff
@@ -312,6 +316,126 @@ impl<W: Write> Write for WriterHasher<W> {
   }
 }
 
+impl<AW: ?Sized + AsyncWrite + Unpin> AsyncWrite for WriterHasher<&mut AW> {
+  fn poll_write(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+    buf: &[u8],
+  ) -> Poll<Result<usize, Error>> {
+    let inner = Pin::new(&mut *self.inner);
+    let result = inner.poll_write(cx, buf);
+    if let Poll::Ready(Ok(written)) = result {
+      // Hash the bytes that were successfully written.
+      self.hasher.update(&buf[0..written]);
+      self.byte_count += written;
+    }
+    result
+  }
+
+  fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+    let inner = Pin::new(&mut *self.inner);
+    inner.poll_flush(cx)
+  }
+
+  fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+    let inner = Pin::new(&mut *self.inner);
+    inner.poll_shutdown(cx)
+  }
+}
+
+///
+/// Copy the data from reader and hash the bytes in one pass.
+/// Use hash() to just hash without copying the data anywhere.
+///
+pub fn copy_and_hash<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<Digest>
```

Comment:
What do you think about naming these sync ones explicitly, e.g.

```suggestion
pub fn sync_copy_and_hash<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<Digest>
```

since, I imagine, the async versions should be (strongly) preferred, except for the LMDB code.

Additionally, it looks like there's potentially a few of these sync ones (`hash`, `hash_path`) that aren't used (either here or #18153); maybe they could disappear, or be switched to `async`?